File: backend/api/models/collaboration_request.py
from django.db import models
from .project import Project
from django.utils import timezone
from api.models.collaboration_invite import CollaborationInvite


class CollaborationRequest(models.Model):
    from_user = models.ForeignKey(
        'auth.User',
        on_delete=models.CASCADE,
        related_name="collab_request_sent",
    )
    to_user = models.ForeignKey(
        'auth.User',
        on_delete=models.CASCADE,
        related_name="collab_request_received",
    )
    to_project = models.ForeignKey(
        Project,
        on_delete=models.CASCADE,
        related_name="collab_request_of_project"
    )
    message = models.CharField(max_length=200, blank=True, default='')
    created = models.DateTimeField(blank=True, null=True)
    rejected = models.DateTimeField(blank=True, null=True)

    class Meta:
        ordering = ['created']

    def accept(self):
        self.to_project.members.add(self.from_user)
        self.delete()
        CollaborationInvite.objects.filter(
            from_user=self.to_user,
            to_user=self.from_user,
            to_project=self.to_project
        ).delete()
        # TODO add notification signal

    def reject(self):
        self.rejected = timezone.now()
        self.save()
        # TODO add notification signal

File: android/app/src/main/java/com/bounswe2020group3/paperlayer/home/data/CollaborationRequest.kt
package com.bounswe2020group3.paperlayer.home.data

import com.squareup.moshi.Json

data class CollaborationRequest(
        @field:Json(name = "id")
        var id: Int,
        @field:Json(name = "to_user")
        var to_user: String,
        @field:Json(name = "from_user")
        var from_user: String,
        @field:Json(name = "created")
        var created: String,
        @field:Json(name = "rejected")
        var rejected: String,
        @field:Json(name = "to_project")
        var to_project: Int,
        @field:Json(name = "message")
        var message: String,
)

